academy/ibc-dev/ibcmw-develop.md

```diff
@@ -0,0 +1,484 @@
+
+Check out the code snippets below to see this in action, or use the links to see how it is implemented for fee middleware (ICS-29).
+
+<HighlightBox type="info">
+
+Middleware **may** choose not to call the underlying application's callback at all. However, these should generally be limited to error cases.
+
+</HighlightBox>
+
+### Middleware version negotiation
+
+In the case where the IBC middleware expects to speak to a compatible IBC middleware on the counterparty chain, they must **use the channel handshake to negotiate the middleware version without interfering in the version negotiation of the underlying application**.
+
+Middleware accomplishes this by formatting the version in a JSON-encoded string containing the middleware version and the application version. The application version may as well be a JSON-encoded string, possibly including further middleware and app versions if the application stack consists of multiple middlewares wrapping a base application. The format of the version is specified in ICS-30 as follows:
+
+```json
+{
+  "<middleware_version_key>": "<middleware_version_value>",
+  "app_version": "<application_version_value>"
+  // ... other custom parameter fields
+}
+```
+
+The `<middleware_version_key>` key in the JSON struct should be replaced by the actual name of the key for the corresponding middleware (e.g. `fee_version`).
+
+During the handshake callbacks, the middleware can unmarshal the version string and retrieve the middleware and application versions. It can do its negotiation logic on `<middleware_version_value>`, and pass the `<application_version_value>` to the underlying application.
+
+<HighlightBox type="note">
+
+Middleware that does not need to negotiate with a counterparty middleware on the remote stack will not implement the version unmarshalling and negotiation, and may simply perform its own custom logic on the callbacks without relying on the counterparty behaving similarly.
+
+</HighlightBox>
+
+### Capabilities
+
+The middleware should simply pass the capability in the callback arguments along to the underlying application so that it may be claimed by the base application. The base application will then pass the capability up the stack in order to authenticate an outgoing packet/acknowledgment.
+
+### Code snippets
+
+Note that the code snippets below contain _pseudo code_, like `doCustomLogic(args)`. Every code snippet is accompanied by a reference to the respective function in the [fee middleware IBC application](https://github.com/cosmos/ibc-go/tree/main/modules/apps/29-fee) from the official `ibc-go` repository.
+
+<ExpansionPanel title="`OnChanOpenInit`">
```

Comment:
We currently have an issue on the platform with rendering inline code as title in these panels (it's not converted to an inline code block), I'll add a feature request for that to the platform, so we can keep these like this.